homeassistant/components/directv/config_flow.py

```diff
@@ -0,0 +1,146 @@
+
+
+def get_ip(host: str) -> str:
+    """Translate hostname to IP address."""
+    if host is None:
+        return None  # pragma: no cover
+
+    return socket.gethostbyname(host)
+
+
+def get_dtv_version(host: str, port: int = DEFAULT_PORT) -> Any:
+    """Test the device connection by retrieving the receiver version info."""
+    # directpy does IO in constructor.
+    dtv = DIRECTV(host, port)
+    return dtv.get_version()
+
+
+async def validate_input(hass: HomeAssistantType, data: Dict) -> Dict:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    host = await hass.async_add_executor_job(get_ip, data["host"])
+    version_info = await hass.async_add_executor_job(get_dtv_version, host)
+
+    return {
+        "title": host,
+        "host": host,
+        "receiver_id": "".join(version_info["receiverId"].split()),
+    }
+
+
+class DirecTVConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for DirecTV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    async def _show_form(self, errors: Optional[Dict] = None) -> Dict[str, Any]:
+        """Show the form to the user."""
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors or {},
+        )
+
+    async def async_step_import(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by yaml file."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by user."""
+        errors = {}
+
+        if not user_input:
+            return await self._show_form()
+
+        try:
+            info = await validate_input(self.hass, user_input)
+            user_input[CONF_HOST] = info[CONF_HOST]
+        except (OSError, RequestException):
+            errors["base"] = ERROR_CANNOT_CONNECT
+            return await self._show_form(errors)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = ERROR_UNKNOWN
+            return await self._show_form(errors)
+
+        await self.async_set_unique_id(info["receiver_id"])
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(title=info["title"], data=user_input)
+
+    async def async_step_ssdp(
+        self, discovery_info: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by discovery."""
+        host = urlparse(discovery_info[ATTR_SSDP_LOCATION]).hostname
+        host = await self.hass.async_add_executor_job(get_ip, host)
+        receiver_id = discovery_info[ATTR_UPNP_SERIAL][4:]  # strips off RID-
+
+        await self.async_set_unique_id(receiver_id)
+        self._abort_if_unique_id_configured()
```

Comment:
This method takes an argument with updates for the config entry data if it is already configured. It might be beneficial to pass `{CONF_HOST: host}` so that if the receiver gets a new IP address, it automatically recovers.